WIP: Retrofit converter and sealed class for response types

akwc10 opened the pull request.

akwc10 committed:
WIP Retrofit custom adapter

akwc10 committed:
Added ResponseCodeInterceptor, removed custom adapter

akwc10 committed:
Sealed class of response types, convert errorBody response and more

Sealed class for 3 restricted class hierarchies.
Converter to convert errorBody response into correct response type.
String extension to count occurrences of matching string.
Manifest correction to lock portrait orientation.
Gradle and retrofit version updates.

fsbarata commented on app/src/main/java/com/my/journeyplanner/helpers/JourneyPlannerApiService.kt:
```diff
@@ -13,16 +13,21 @@ interface JourneyPlannerApiService {
+            .build()
 
-        fun createApiService(): JourneyPlannerApiService = Retrofit.Builder()
+        val retrofit: Retrofit = Retrofit.Builder()
```
is this change necessary? what have we improved here? Specially concerned about making it public

akwc10 replied:
Used to access JourneyPlannerApiService.retrofit.responseBodyConverter().
I can use fakeRetrofit:
        val fakeRetrofit: Retrofit = Retrofit.Builder()
            .baseUrl("https://ttttt.com/")
            .client(OkHttpClient().newBuilder().build())
            .addConverterFactory(MoshiConverterFactory.create())
            .build()
Am I missing the point?

akwc10 replied:
Renamed to MockRetrofit using MockInterceptor() for responseBodyConverter(). I'd moved this into a separate class but I suspect you have a better solution!

akwc10 replied:
Removed as no longer required

fsbarata commented on app/src/main/java/com/my/journeyplanner/helpers/ResponseCodeInterceptor.kt:
```diff
@@ -0,0 +1,14 @@
+import okhttp3.Interceptor
+import okhttp3.Response
+
+class ResponseCodeInterceptor : Interceptor {
```
Better call it something we know what it is. Say like, MultiOptionResponseCodeInterceptor or something

akwc10 replied:
Updated as advised

fsbarata commented on app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt:
```diff
@@ -20,30 +24,91 @@ class MainPresenter(private val view: MainContract.View) : MainContract.Presente
+                }
             }
 
             override fun onResponse(
```
this method is wayyyy too big. Start breaking it down

akwc10 replied:
De-duped and functionalised

akwc10 replied:
Logic no longer required due to implementation of MultiOptionResponseCodeInterceptor() and JourneyPlannerResultConverterFactory

fsbarata commented on app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt:
```diff
@@ -20,30 +24,91 @@ class MainPresenter(private val view: MainContract.View) : MainContract.Presente
+                    Log.d(TAG, "*** ItineraryResult: ${response.body()} ***")
+//                TODO("Display data in new activity")
+                } else when (response.code()) {
+                    300 -> {
```
didn't we make sure this doesn't happen?

akwc10 replied:
Yes, but I removed the MultiOptionResponseCodeInterceptor() as I couldn't get it to work with:
public @Nullable Converter<ResponseBody, ?> responseBodyConverter(Type type,
        Annotation[] annotations, Retrofit retrofit)
due to the generic return type ResponseBody i.e. I can't do this:
                                val converter: Converter<T, JourneyPlanner.NotIdentifiedResult> =
                                    JourneyPlannerApiService.fakeRetrofit.responseBodyConverter(
                                        JourneyPlanner.NotIdentifiedResult::class.java,
                                        arrayOfNulls<Annotation>(0)
                                    )
  converter.convert(response.body())
where:
  public @Nullable T body() {}
There'll be an obvious way but I haven't found it

fsbarata replied:
yeah I'll help you out on this later

akwc10 replied:
Reinstated MultiOptionResponseCodeInterceptor()

fsbarata commented on app/src/main/java/com/my/journeyplanner/helpers/StringUtils.kt:
```diff
@@ -0,0 +1,12 @@
+    val thisStringIndexIncremented =
+        if (count > 0) index + thisStringIndex + searchedString.length else this.length
+
+    return count + countOccurrences(searchedString, thisStringIndexIncremented)
```
or you can scrap all this code and simply:

fun String.countOccurrences(searchedString: String) = split(searchString).length - 1

Also write unit tests before you refactor.

akwc10 replied:
Feedback implemented

fsbarata commented on app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt:
```diff
@@ -20,30 +24,91 @@ class MainPresenter(private val view: MainContract.View) : MainContract.Presente
-            override fun onFailure(call: Call<JourneyPlannerDisambiguationResult>, throwable: Throwable) {
-                Log.e(TAG, "Other error: ${throwable.message}")
-                Log.e(TAG, "Other error: ${throwable.stackTrace.forEach { println(it) }}")
+        call!!.enqueue(object : Callback<JourneyPlanner.ItineraryResult> {
```
presenters are about usecase->view interaction.

You're mixing the usecase layer here. Better is to ask for data to the usecase, and the use case deals with the api and all its particularities. Presenters just care about wiring logic

akwc10 replied:
Added uses cases but I'm not convinced with my implementation

fsbarata replied:
No, you need something of the sort of a Repository / Provider / etc. I'll give you some ideas of what to do later

akwc10 replied:
Added repository, datasource and use cases/interactors.

akwc10 replied:
Fixed repository which was just delegating to datasource and removed latter

akwc10 committed:
Rename interceptor, simplify String.countOccurrences(), add unit tests

Add fakeRetrofit property for responseBodyConverter().
Gradle-wrapper version update.

akwc10 committed:
Cleanup onResponse() and logging output

akwc10 committed:
MockRetrofit using MockInterceptor() for responseBodyConverter()

Should not be making erroneous calls.

akwc10 committed:
Use cases to separate retrofit framework

akwc10 committed:
Second attempt at clean architecture

Removed incorrect first attempt of use cases.
Added core module containing data, interactors and domain packages.
Added datasource interface and repository to data package.
Moved model to domain package.
Added interactor implementations.
Added framework package to app.
Added interactors data class and datasource implementation to framework package.
Instantiated interactors and datasource in MainActivity where the former is passed as a parameter to MainPresenter.
Cleaned up MainPresenter by moving logic to datasource implementation and replacing with interactors.

akwc10 committed:
Converter factory WIP and various cleanup

Removed MockInterceptor and related artefacts.
Removed unnecessary datasource.
Fixed repository so it no longer delegates and fixed interactors.
Extra result types.

akwc10 committed:
Converter factory implementation

Added converter factories for 5 result types.
Added JSON operations for parsing response body so converter factory type can be determined.

akwc10 committed:
Correction

akwc10 committed:
Fixed typo, json operations are now type extensions, logging improvements

fsbarata commented on app/src/main/java/com/my/journeyplanner/framework/JourneyPlannerResultConverterFactory.kt:
```diff
@@ -0,0 +1,113 @@
+                hasFromLocationDisambiguationOptions && hasToLocationDisambiguationOptions
+            val isFromDisambiguationOptions =
+                hasFromLocationDisambiguationOptions && !hasToLocationDisambiguationOptions
+            val isToDisambiguationOptions =
```
from what I gathered, there was a field called type which had a full qualified type. What happened to that and why did you decide to understand what the data was comprised of?

akwc10 replied:
ItineraryResult i.e. exact match
{"$type":"Tfl.Api.Presentation.Entities.JourneyPlanner.ItineraryResult, Tfl.Api.Presentation.Entities",...
notidentified i.e. no match either way
{"$type":"Tfl.Api.Presentation.Entities.JourneyPlanner.DisambiguationResult, Tfl.Api.Presentation.Entities",...
From To Disambiguation i.e. multiple options both ways
{"$type":"Tfl.Api.Presentation.Entities.JourneyPlanner.DisambiguationResult, Tfl.Api.Presentation.Entities",...
From DisambiguationResult i.e. mutliple options in from direction
{"$type":"Tfl.Api.Presentation.Entities.JourneyPlanner.DisambiguationResult, Tfl.Api.Presentation.Entities",...
To DisambiguationResult i.e. multiple options in to direction
{"$type":"Tfl.Api.Presentation.Entities.JourneyPlanner.DisambiguationResult, Tfl.Api.Presentation.Entities",...

The type is the same for the other types with the exception of ItineraryResult hence why I have to drill down a bit more

akwc10 replied:
Reduced data types from 5 to 2 and removed associated logic

fsbarata commented on app/src/main/java/com/my/journeyplanner/framework/JsonOperations.kt:
```diff
@@ -0,0 +1,59 @@
+                    DISAMBIGUATION_OPTIONS -> {
+                        if (this.getJSONObject(locationDisambiguation).getJSONArray(
+                                key
+                            ).length() > 0
```
isNotEmpty()

akwc10 replied:
JSONArray doesn't implement Collection hence can't apply this

fsbarata commented on app/src/main/java/com/my/journeyplanner/framework/JsonOperations.kt:
```diff
@@ -0,0 +1,59 @@
+            }
+        }
+    } catch (e: JSONException) {
+        Log.e(TAG, e.message)
```
not cool. Why can it go wrong?

akwc10 replied:
It shouldn't throw an exception as I've already checked the keys exists beforehand. Just being overcautious. Removed

fsbarata commented on app/src/main/java/com/my/journeyplanner/framework/JsonOperations.kt:
```diff
@@ -0,0 +1,59 @@
+                    }
+                }
+            } else {
+                Log.d(TAG, "Missing key: $locationDisambiguation.$key")
```
what's the problem with this?

akwc10 replied:
Just in case TFL decide to change the API. Remove it?

akwc10 committed:
Reduced data classes from 5 to 2 and removed associated logic
